src/oidcendpoint/session/grant.py

```diff
@@ -0,0 +1,355 @@
+                token_handler = endpoint_context.session_manager.token_handler.handler[
+                    GRANT_TYPE_MAP[token_type]]
+
+            item.value = token_handler(session_id=session_id,
+                                       **self.payload_arguments(session_id,
+                                                                endpoint_context,
+                                                                token_type=token_type,
+                                                                scope=scope))
+        else:
+            raise ValueError("Can not mint that kind of token")
+
+        self.issued_token.append(item)
+        self.used += 1
+        return item
+
+    def get_token(self, value: str) -> Optional[Token]:
+        for t in self.issued_token:
+            if t.value == value:
+                return t
+        return None
+
+    def revoke_token(self,
+                     value: Optional[str] = "",
+                     based_on: Optional[str] = "",
+                     recursive: bool = True):
+        for t in self.issued_token:
+            if not value and not based_on:
+                t.revoked = True
+            elif value and based_on:
+                if value == t.value and based_on == t.based_on:
+                    t.revoked = True
+            elif value and t.value == value:
+                t.revoked = True
+                if recursive:
+                    self.revoke_token(based_on=t.value)
+            elif based_on and t.based_on == based_on:
+                t.revoked = True
+                if recursive:
+                    self.revoke_token(based_on=t.value)
+
+    def get_spec(self, token: Token) -> Optional[dict]:
+        if self.is_active() is False or token.is_active is False:
+            return None
+
+        res = {}
+        for attr in ["scope", "claims", "resources"]:
+            _val = getattr(token, attr)
+            if _val:
+                res[attr] = _val
+            else:
+                _val = getattr(self, attr)
+                if _val:
+                    res[attr] = _val
+        return res
+
+
+DEFAULT_USAGE = {
+    "authorization_code": {
+        "max_usage": 1,
+        "supports_minting": ["access_token", "refresh_token", "id_token"],
+        "expires_in": 300
+    },
+    "access_token": {
+        "supports_minting": [],
+        "expires_in": 3600
+    },
+    "refresh_token": {
+        "supports_minting": ["access_token", "refresh_token", "id_token"]
+    }
+}
+
+
+def get_usage_rules(token_type, endpoint_context, grant, client_id):
+    """
+    The order of importance:
+    Grant, Client, EndPointContext, Default
+
+    :param token_type: The type of token
+    :param endpoint_context: An EndpointContext instance
+    :param grant: A Grant instance
+    :param client_id: The client identifier
+    :return: Usage specification
+    """
+
+    _usage = endpoint_context.authz.usage_rules_for(token_type)
```

Comment:
```python
     _usage = endpoint_context.authz.usage_rules_for(client_id, token_type)
```

Reply:
Hmm, tests should have caught this. Have to write more.